csharp/src/Apache.Arrow/Arrays/ArrayDataConcatenator.cs

```diff
@@ -0,0 +1,240 @@
+
+                Result = new ArrayData(type, _totalLength, _totalNullCount, 0, new ArrowBuffer[] { validityBuffer, offsetBuffer, valueBuffer });
+            }
+
+            private ArrowBuffer ConcatenateValidityBuffer()
+            {
+                if (_totalNullCount == 0)
+                {
+                    return ArrowBuffer.Empty;
+                }
+
+                return ConcatenateBitmapBuffer(0);
+            }
+
+            private ArrowBuffer ConcatenateBitmapBuffer(int bufferIndex)
+            {
+                var builder = new ArrowBuffer.BitmapBuilder(_totalLength);
+
+                foreach (ArrayData arrayData in _arrayDataList)
+                {
+                    int length = arrayData.Length;
+                    ReadOnlySpan<byte> span = arrayData.Buffers[bufferIndex].Span;
+
+                    for (int i = 0; i < length; i++)
+                    {
+                        builder.Append(span.IsEmpty || BitUtility.GetBit(span, i));
+                    }
+                }
+
+                return builder.Build(_allocator);
+            }
+
+            private ArrowBuffer ConcatenateFixedWidthTypeValueBuffer(FixedWidthType type)
+            {
+                int typeByteWidth = type.BitWidth / 8;
+                var builder = new ArrowBuffer.Builder<byte>(_totalLength * typeByteWidth);
+
+                foreach (ArrayData arrayData in _arrayDataList)
+                {
+                    int length = arrayData.Length;
+                    int byteLength = length * typeByteWidth;
+
+                    builder.Append(arrayData.Buffers[1].Span.Slice(0, byteLength));
+                }
+
+                return builder.Build(_allocator);
+            }
+
+            private ArrowBuffer ConcatenateVariableBinaryValueBuffer()
+            {
+                var builder = new ArrowBuffer.Builder<byte>();
+
+                foreach (ArrayData arrayData in _arrayDataList)
+                {
+                    int lastOffset = arrayData.Buffers[1].Span.CastTo<int>()[arrayData.Length];
+                    builder.Append(arrayData.Buffers[2].Span.Slice(0, lastOffset));
+                }
+
+                return builder.Build(_allocator);
+            }
+
+            private ArrowBuffer ConcateneteOffsetBuffer()
+            {
+                var builder = new ArrowBuffer.Builder<int>(_totalLength + 1);
+                int baseOffset = 0;
+
+                builder.Append(0);
+
+                foreach (ArrayData arrayData in _arrayDataList)
+                {
+                    if (arrayData.Length == 0)
+                    {
+                        continue;
+                    }
+
+                    // The first offset is always 0.
+                    // It should be skipped because it duplicate to the last offset of builder.
+                    ReadOnlySpan<int> span = arrayData.Buffers[1].Span.CastTo<int>().Slice(1, arrayData.Length);
```

Comment:
Doesn't this need to be `arrayData.Length - 1`? Because you sliced off the first element.
```suggestion
                    ReadOnlySpan<int> span = arrayData.Buffers[1].Span.CastTo<int>().Slice(1, arrayData.Length - 1);
```

Reply:
It is a bit confusing, but the length of `arrayData.Buffers[1]` is `arrayData.Length + 1`, so I think being `arrayData.Length` is correct.